clientconn.go

```diff
@@ -919,6 +909,25 @@ func (ac *addrConn) tryUpdateAddrs(addrs []resolver.Address) bool {
 	return curAddrFound
 }
 
+// getServerName determines the serverName to be used in the connection
+// handshake. The default value for the serverName is the authority on the
+// ClientConn, which either comes from the user's dial target or through an
+// authority override specified using the WithAuthority dial option. Name
+// resolvers can specify a per-address override for the serverName through the
+// resolver.Address.ServerName field which is used only if the WithAuthority
+// dial option was not used. The rationale is that per-address authority
+// overrides specified by the name resolver can represent a security risk, while
+// an override specified by the user is more dependable since they probably know
+// what they are doing.
+func (ac *addrConn) getServerName(addr resolver.Address) string {
+	if ac.cc.dopts.authority == "" {
+		if addr.ServerName != "" {
+			return addr.ServerName
+		}
+	}
```

Comment:
How about this, which more clearly shows the order of precedence:

```go
if ac.cc.dopts.authority != "" {
	return ac.cc.dopts.authority
}
if addr.ServerName != "" {
	return addr.ServerName
}
return ac.cc.authority
```

(And delete the above code that sets `cc.authority` from the dial option, as mentioned above.)